### PR TITLE
Get, Update, Post and delete catalog promotions

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         | master 
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

extends https://github.com/Sylius/Sylius/pull/12986

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/CatalogPromotion.xml [hunk 1]
```diff
@@ -36,6 +36,26 @@
...
+            <itemOperation name="admin_get">
+                <attribute name="method">GET</attribute>
+                <attribute name="normalization_context">
+                    <attribute name="groups">admin:catalog_promotion:read</attribute>
+                </attribute>
+            </itemOperation>
+
+            <itemOperation name="admin_put">
+                <attribute name="method">PUT</attribute>
+                <attribute name="denormalization_context">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/CatalogPromotion.xml:49]: You should add also a normalization context here and in admin_post

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 2]
```diff
@@ -43,4 +43,121 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+
+    /** @test */
+    public function it_allows_admin_to_get_catalog_promotion(): void
+    {
+        $catalogPromotion = $this->loadFixturesFromFiles(['authentication/api_administrator.yaml', 'catalog_promotion.yaml'])['catalog_promotion'];
+
+        $token = $this->logInAdminUser('api@example.com');
+        $authorizationHeader = self::$container->getParameter('sylius.api.authorization_header');
+        $header['HTTP_' . $authorizationHeader] = 'Bearer ' . $token;
+        $header = array_merge($header, self::CONTENT_TYPE_HEADER);
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:55]: Maybe you can extract these lines as they are needed for every test

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 3]
```diff
@@ -43,4 +43,121 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+
+        $this->client->request(
+            'POST',
+            '/api/v2/admin/catalog-promotions',
+            [],
+            [],
+            $header,
+            json_encode(["name" => "new_promotion", "code" => "new_code"], JSON_THROW_ON_ERROR)
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:86]:
```suggestion
            json_encode(['name' => 'new_promotion', 'code' => 'new_code'], JSON_THROW_ON_ERROR)
```

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 4]
```diff
@@ -43,4 +43,121 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+            $header
+        );
+
+        $response = $this->client->getResponse();
+
+        $this->assertResponse($response, 'admin/get_catalog_promotion_admin_response', Response::HTTP_OK);
+    }
+
+    /** @test */
+    public function it_allows_admin_to_post_new_catalog_promotion(): void
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:71]:
```suggestion
    public function it_allows_admin_to_create_a_catalog_promotion(): void
```

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 5]
```diff
@@ -43,4 +43,121 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+        $response = $this->client->getResponse();
+
+        $this->assertResponseCode($response, Response::HTTP_CREATED);
+    }
+
+    /** @test */
+    public function it_allows_admin_to_update_catalog_promotion(): void
+    {
+        $this->loadFixturesFromFile('authentication/api_administrator.yaml');
+        $this->loadFixturesFromFile('catalog_promotion.yaml');
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:98]:
```suggestion
        $this->loadFixturesFromFiles(['authentication/api_administrator.yaml', 'catalog_promotion.yaml']);
```

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 6]
```diff
@@ -43,4 +43,121 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+            'GET',
+            '/api/v2/admin/catalog-promotions',
+            [],
+            [],
+            $header
+        );
+
+        $response = $this->client->getResponse();
+        $content = json_decode($response->getContent(), true);
+        return (string) $content['hydra:member'][0]['id'];
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:161]:
```suggestion

        return (string) $content['hydra:member'][0]['id'];
```

**Review by GSadee — COMMENTED** [review]
IMO, in this PR you should add also behat scenarios for updating and deleting catalog promotion

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 7]
```diff
@@ -43,4 +43,121 @@ public function it_allows_admin_to_get_catalog_promotions(): void
 
         $this->assertResponse($response, 'admin/get_catalog_promotions_admin_response', Response::HTTP_OK);
     }
+
+    /** @test */
+    public function it_allows_admin_to_get_catalog_promotion(): void
+    {
+        $catalogPromotion = $this->loadFixturesFromFiles(['authentication/api_administrator.yaml', 'catalog_promotion.yaml'])['catalog_promotion'];
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:50]:
```suggestion
        $fixtures = $this->loadFixturesFromFiles(['authentication/api_administrator.yaml', 'catalog_promotion.yaml']);
```

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 8]
```diff
@@ -43,4 +43,121 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+
+        $token = $this->logInAdminUser('api@example.com');
+        $authorizationHeader = self::$container->getParameter('sylius.api.authorization_header');
+        $header['HTTP_' . $authorizationHeader] = 'Bearer ' . $token;
+        $header = array_merge($header, self::CONTENT_TYPE_HEADER);
+
+        $this->client->request(
+            'GET',
+            sprintf('/api/v2/admin/catalog-promotions/%s', $catalogPromotion['id']),
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:59]: You should use here a code instead of id, as the code is an identifier of catalog promotion entity
```suggestion
            sprintf('/api/v2/admin/catalog-promotions/%s', $fixtures['catalog_promotion']['code']),
```

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 9]
```diff
@@ -43,4 +43,121 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+        $this->loadFixturesFromFile('authentication/api_administrator.yaml');
+        $this->loadFixturesFromFile('catalog_promotion.yaml');
+
+        $token = $this->logInAdminUser('api@example.com');
+        $authorizationHeader = self::$container->getParameter('sylius.api.authorization_header');
+        $header['HTTP_' . $authorizationHeader] = 'Bearer ' . $token;
+        $header = array_merge($header, self::CONTENT_TYPE_HEADER);
+
+        // Should be replaced with fixture_catalog_promotion['id'];
+        $id = $this->getId($header);
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:106]: Use code instead of id and remove this unnecessary method

**Comment by arti0090** [comment]
Behat tests for this feature will be added in next PR 🎉

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/CatalogPromotion.xml [hunk 10]
```diff
@@ -30,12 +30,38 @@
...
+                <attribute name="normalization_context">
+                    <attribute name="groups">admin:catalog_promotion:read</attribute>
+                </attribute>
+            </itemOperation>
+
+            <itemOperation name="admin_put">
+                <attribute name="method">PUT</attribute>
+                <attribute name="normalization_context">
+                    <attribute name="groups">admin:catalog_promotion:update</attribute>
+                </attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/CatalogPromotion.xml:54]:
```suggestion
                <attribute name="normalization_context">
                    <attribute name="groups">admin:catalog_promotion:read</attribute>
                </attribute>
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/CatalogPromotion.xml [hunk 11]
```diff
@@ -30,12 +30,38 @@
 
             <collectionOperation name="admin_post">
                 <attribute name="method">POST</attribute>
+                <attribute name="normalization_context">
+                    <attribute name="groups">admin:catalog_promotion:create</attribute>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/CatalogPromotion.xml:34]:
```suggestion
                    <attribute name="groups">admin:catalog_promotion:read</attribute>
```

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 12]
```diff
@@ -25,6 +26,26 @@ final class CatalogPromotionsTest extends JsonApiTestCase
...
+            'GET',
+            '/api/v2/admin/catalog-promotions',
+            [],
+            [],
+            $header
+        );
+
+        $response = $this->client->getResponse();
+
+        $this->assertResponse($response, 'admin/catalog_promotion/get_catalog_promotions_admin_response', Response::HTTP_OK);
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:41]:
```suggestion
        $this->assertResponse($response, 'admin/catalog_promotion/get_catalog_promotions_response', Response::HTTP_OK);
```

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 13]
```diff
@@ -33,14 +54,91 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+            sprintf('/api/v2/admin/catalog-promotions/%s', $catalogPromotion->getCode()),
             [],
             [],
             $header
         );
 
         $response = $this->client->getResponse();
 
-        $this->assertResponse($response, 'admin/get_catalog_promotions_response', Response::HTTP_OK);
+        $this->assertResponse($response, 'admin/catalog_promotion/get_catalog_promotion_admin_response', Response::HTTP_OK);
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:65]:
```suggestion
        $this->assertResponse($response, 'admin/catalog_promotion/get_catalog_promotion_response', Response::HTTP_OK);
```

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 14]
```diff
@@ -33,14 +54,91 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+            '/api/v2/admin/catalog-promotions',
+            [],
+            [],
+            $header,
+            json_encode(['name' => 'new_promotion', 'code' => 'new_code'], JSON_THROW_ON_ERROR)
+        );
+
+        $response = $this->client->getResponse();
+
+        $this->assertResponse($response, 'admin/catalog_promotion/post_catalog_promotion_admin_response', Response::HTTP_CREATED);
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:85]:
```suggestion
        $this->assertResponse($response, 'admin/catalog_promotion/post_catalog_promotion_response', Response::HTTP_CREATED);
```

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 15]
```diff
@@ -33,14 +54,91 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+            sprintf('/api/v2/admin/catalog-promotions/%s', $catalogPromotion->getCode()),
+            [],
+            [],
+            $header,
+            json_encode(['name' => 'new_promotion', 'code' => 'new_code'], JSON_THROW_ON_ERROR)
+        );
+
+        $response = $this->client->getResponse();
+
+        $this->assertResponse($response, 'admin/catalog_promotion/put_catalog_promotion_admin_response', Response::HTTP_OK);
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:105]:
```suggestion
        $this->assertResponse($response, 'admin/catalog_promotion/put_catalog_promotion_response', Response::HTTP_OK);
```

File: tests/Api/Admin/CatalogPromotionsTest.php [hunk 16]
```diff
@@ -33,14 +54,91 @@ public function it_allows_admin_to_get_catalog_promotions(): void
...
+
+        return $catalogPromotion;
+    }
+
+    private function getLoggedHeader(): array
+    {
+        $token = $this->logInAdminUser('api@example.com');
+        $authorizationHeader = self::$container->getParameter('sylius.api.authorization_header');
+        $header['HTTP_' . $authorizationHeader] = 'Bearer ' . $token;
+        return array_merge($header, self::CONTENT_TYPE_HEADER);
```
Comment (GSadee) [inline tests/Api/Admin/CatalogPromotionsTest.php:142]:
```suggestion

        return array_merge($header, self::CONTENT_TYPE_HEADER);
```

File: tests/Api/Responses/Expected/admin/catalog_promotion/get_catalog_promotions_admin_response.json [hunk 17]
```diff
@@ -6,7 +6,7 @@
         {
             "@id": "\/api\/v2\/admin\/catalog-promotions\/mugs_discount",
             "@type": "CatalogPromotion",
-            "id": @integer@,
+            "id": "@integer@",
```
Comment (GSadee) [inline tests/Api/Responses/Expected/admin/catalog_promotion/get_catalog_promotions_admin_response.json:9]: Why did you change that? IMO here and in other responses there should be:
```suggestion
            "id": @integer@,
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, @arti0090! :1st_place_medal:
